File: kt/main/kotlin/pushtofunction/fs/fooSfl.kt
/*
 *  Copyright © 2022 Paulo Villela. All rights reserved.
 *  Use of this source code is governed by the MIT license
 *  that can be found in the LICENSE file.
 */

package pushtofunction.fs

data class FooSflCfgInfo(
	val x: String
)

data class FooSflCfgSrc(
	val get: () -> FooSflCfgInfo,
	val bar: BarBfT
)

typealias FooSflT = () -> Unit

fun fooSflC(cfg: FooSflCfgSrc): FooSflT {
	val barBf = cfg.bar
	return {
		println(cfg.get().x)
		barBf()
	}
}

File: ts/src/push-to-function/fs/foo-sfl.ts
/*
 * Copyright © 2022 Paulo Villela. All rights reserved.
 * Use of this source code is governed by the MIT license
 * that can be found in the LICENSE file.
 */

export type FooSflCfgInfo = {
  x: string;
};

export type FooSflCfgSrc = {
  get: () => FooSflCfgInfo;
  bar: () => void;
};

export type FooSflT = () => void;

export function fooSflC(cfg: FooSflCfgSrc): FooSflT {
  return function() {
    console.log(cfg.get().x);
    cfg.bar()
  }
}
